packages/eth-json-rpc-middleware/src/block-cache.test.ts

```diff
@@ -1,57 +1,373 @@
-import { PollingBlockTracker } from '@metamask/eth-block-tracker';
-import { providerFromEngine } from '@metamask/eth-json-rpc-provider';
 import { JsonRpcEngine } from '@metamask/json-rpc-engine';
-import pify from 'pify';
+import type { Json, JsonRpcSuccess } from '@metamask/utils';
 
 import { createBlockCacheMiddleware } from '.';
-import createHitTrackerMiddleware from '../test/util/createHitTrackerMiddleware';
+import {
+  createProviderAndBlockTracker,
+  createRequest,
+  stubProviderRequests,
+} from '../test/util/helpers';
 
-function createTestSetup() {
-  // raw data source
-  // create block tracker
-  // create higher level
-  const engine = new JsonRpcEngine();
-  const provider = providerFromEngine(engine);
+describe('block cache middleware', () => {
+  let provider: ReturnType<typeof createProviderAndBlockTracker>['provider'];
+  let blockTracker: ReturnType<
+    typeof createProviderAndBlockTracker
+  >['blockTracker'];
 
-  const blockTracker = new PollingBlockTracker({
-    provider,
+  beforeEach(() => {
```

Comment:
Any reason for preferring `beforeEach` over factory functions in tests? In my experience, they serve as a sort of junk drawer for setup data that has a tendency to get more and more unreadable over time, and they make it more difficult to write tests that extend the setup data in some way. (We have some pretty egregious examples of how not to use `beforeEach` in existing controller packages.) But I also realize that there are some advantages to using them.

Reply:
I generally agree with you, except in the case where there's cleanup to perform, in this case `blockTracker.destroy()`. In such cases I generally find the hooks to be worth it.